flashlight/fl/tensor/backend/jit/JitBackend.cpp

```diff
@@ -8,19 +8,253 @@
+    oldToNewAxes.resize(inputRank);
+    std::reverse(outputDims.begin(), outputDims.end());
+    std::iota(oldToNewAxes.begin(), oldToNewAxes.end(), 0);
+    std::reverse(oldToNewAxes.begin(), oldToNewAxes.end());
+  } else if (axes.ndim() == inputRank) {
+    for (int axis = 0; axis < axes.ndim(); axis++) {
+      outputDims[axis] = inputShape.dim(oldToNewAxes[axis]);
+    }
+  } else {
+    std::ostringstream oss;
+    oss << "[JitBackend::transpose] Invalid axes: " << axes
+        << " for shape: " << inputShape;
+    throw std::runtime_error(oss.str());
+  }
+  return Shape(outputDims);
+}
+
+Shape inferTileOutputShape(const Shape& inputShape, const Shape& tileDims) {
+  // TODO refactor with logic in OneDnnBackend
+  std::vector<Dim> paddedTensorDims = inputShape.get();
+  std::vector<Dim> paddedTileDims = tileDims.get();
+  const auto inputRank = inputShape.ndim();
+  const auto tileRank = tileDims.ndim();
+  if (inputRank > tileRank) {
+    const auto diff = inputRank - tileRank;
+    paddedTileDims.insert(paddedTileDims.end(), diff, 1);
+  } else {
+    const auto diff = tileRank - inputRank;
+    paddedTensorDims.insert(paddedTensorDims.end(), diff, 1);
+  }
+  std::vector<Dim> outputDims;
+  for (unsigned i = 0; i < paddedTensorDims.size(); i++) {
+    outputDims.push_back(paddedTensorDims[i] * paddedTileDims[i]);
+  }
+  return Shape(outputDims);
+}
+
+Shape inferConcatenateOutputShape(
+    const std::vector<Tensor>& tensors,
+    const unsigned axisToConcat) {
+  // TODO need a nice way to construct empty tensor for wrapped backend
+  if (tensors.empty()) {
+    throw std::runtime_error(
+        "[inferConcatenateOutputShape] Nothing to concatenate");
+  }
+  const auto shapes = broadcastShapesToMaxRank(tensors, axisToConcat + 1);
+  const unsigned outputRank = shapes.front().ndim();
+  if (axisToConcat >= outputRank) {
+    std::ostringstream oss;
+    oss << "[inferConcatenateOutputShape] Axis too big: " << axisToConcat;
+    throw std::runtime_error(oss.str());
+  }
+  std::vector<Dim> outputDims;
+  for (unsigned axis = 0; axis < outputRank; axis++) {
+    Dim concatenatedDim = 0;
+    if (axis == axisToConcat) {
+      for (const auto& shape : shapes) {
+        concatenatedDim += shape.dim(axis);
+      }
+    } else {
+      concatenatedDim = shapes.front().dim(axis);
+    }
+    outputDims.push_back(concatenatedDim);
+  }
+  return Shape(outputDims);
+}
+
+Shape inferPadOutputShape(
+    const Shape& inputShape,
+    const std::vector<std::pair<int, int>>& padWidths) {
+  std::vector<Dim> outputDims = inputShape.get();
+  if (padWidths.size() > static_cast<size_t>(inputShape.ndim())) {
+    throw std::runtime_error("[inferPadOutputShape] too many paddings");
+  }
+  for (unsigned axis = 0; axis < padWidths.size(); axis++) {
+    const auto& [beforeDim, afterDim] = padWidths[axis];
+    outputDims[axis] += beforeDim + afterDim;
+  }
+  return Shape(outputDims);
+}
+
+Shape inferMatmulOutputShape(
```

Comment:
Moved them to a `ShapeInference.h` file, and dropped a TODO suggesting we refactor it with AF & OneDNN backend, and allow future backend implementation to access and reuse them.

Regarding looking up shape inference,

1. I'll likely add a `matmul` node since it behaves a bit different from regular binops (batching and transpose).
2. Right now I'm attaching the logic to `Node` subclasses (e.g., inside `BinaryNode::create`). I do know that existing frameworks like TVM has a decentralized way of storing such logic (shape, type inference). But I kinda wanted to avoid having another `ShapeInference` pass that annotates `Node::optionalShape`, basically KISS for now I guess, and see how things evolve:).